lib/teleterm/daemon/daemon.go

```diff
@@ -213,12 +215,78 @@ func (s *Service) createGateway(ctx context.Context, params CreateGatewayParams)
 }
 
 func (s *Service) onExpiredGatewayCert(ctx context.Context, gateway *gateway.Gateway) error {
-	cluster, err := s.ResolveCluster(gateway.TargetURI())
+	certReissuer, err := s.newGatewayCertReissuer(gateway)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	return trace.Wrap(s.cfg.GatewayCertReissuer.ReissueCert(ctx, gateway, cluster))
+	if err := certReissuer.ReissueCert(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// newGatewayCertReissuer creates a new DB cert reissuer.
+// If you want to customize a DBCertReissuer, for example for testing purposes, you
+// may create and configure an instance directly, without calling this method.
+func (s *Service) newGatewayCertReissuer(g *gateway.Gateway) (*gateway.CertReissuer, error) {
+	cluster, err := s.ResolveCluster(g.TargetURI())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &gateway.CertReissuer{
+		Gateway:        g,
+		ReissueDBCerts: cluster.ReissueDBCerts,
+		Relogin:        s.relogin,
+		NotifyError:    s.notifyAppAboutReissueError,
+	}, nil
+}
+
+// relogin makes the Electron app display a login modal to trigger re-login.
+func (s *Service) relogin(ctx context.Context, gateway *gateway.Gateway) error {
+	// The Electron app cannot display two login modals at the same time, so we have to cut short any
+	// concurrent relogin requests.
+	if !s.reloginMu.TryLock() {
```

Comment:
I used `TryLock` in the previous version of this, I wasn't sure if it was a valid use case for `TryLock` but [Tiago said it is](https://gravitational.slack.com/archives/C0DF0TPMY/p1662041810499619?thread_ts=1662040391.272729&cid=C0DF0TPMY). ;)

I used it to manage a scenario where two db clients, connected to separate local proxies via Connect, attempted connection after the proxies' certificates expired. If a db client tries to connect via a proxy with an expired cert, we bring focus to the Electron app and display the login modal.

Should the second client try connecting via the other proxy while the login modal is displayed (and `reloginMu` is locked), we abort the connection instead of waiting for the `reloginMu` unlock. We want to prevent clients from waiting for each other to avoid connection timeouts (which are less clear than cutting the connection attempt ASAP) or cascading modals.

Given that we and users don't fully control db clients, which might try connecting to a db whenever they decide, we chose the option with a clearer failure mode.